### PR TITLE
Update docsprint to reflect the new xapian-letor.

### PR DESCRIPTION
**Pull request by AyushP123**
I have updated the docsprint to account for changes in xapian-letor.

### CODE REVIEW

**Comment by jaylett** [comment]
As I mentioned on IRC, this cannot maintain the structure of the documentation file from `xapian-letor`. I believe it needs to have a _very_ brief introduction into what letor actually does (it reorders a match set to bring more relevant documents to the top, using a more sophisticated approach to calculating relevance than probabilistic weighting schemes like BM25), then show how to use it. Finally, at the end, it can talk about the background in more detail; but you don't have to understand it in order to use letor.

You should look at the rest of the repo to see how examples are done. They should be complete programs, out of which specific line ranges are included in the documentation. You'll have to write the examples in C++ (because language bindings aren't available for letor yet). The way this is done will also provide syntax highlighting for the code snippets.

And you _must_ build documentation and look at it before opening a pull request. This PR has a number of RST syntax errors and incorrect uses which are obvious as soon as you look at the rendered output. This is the documentation equivalent of tidying up C++ source code: it must build without errors, and it should follow the conventions already established in the rest of the documentation, as far as possible.

A final note: all the stuff about extendability should be removed. As it stands, it's either obvious (to add a new `Scorer`, you subclass `Scorer`) or incorrect (users of Xapian don't have to put new scorers into the Xapian build tree, and indeed should not). Once we have basic documentation in place we can perhaps think of a way of explaining with a practical example how to extend letor.

I suggest you make the relevant changes then force push a new commit, and then add a comment here to indicate that it's ready for review; right now there are enough issues you can pick up yourself from the above without it being worth pointing every single one out, and given the need to restructure the document, I think fixup commits are going to be hard to follow.

File: advanced/learning_to_rank.rst [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+Xapian Learning-to-Rank
+=======================
+
+.. contents:: Table of Contents
+
+
+Introduction
+============
+
+Learning-to-Rank(LTR) can be viewed as a weighting scheme which involves machine learning. The main idea behind LTR is to bring up relevant documents given a low ranking by probablistic techniques like BM25 by using machine learning models. A model is trained by learning from the relevance judgements provided by a user corresponding to a set of queries and a corpus of documents. This model is then used to re-rank the matchset to bring more relevant documents higher in the ranking. Learning-to-Rank has gained immense popularity and attention among researchers recently. Xapian is the first project with Learning-to-Rank functionality added to it.
```
Comment (ojwb) [inline advanced/learning_to_rank.rst]: > Xapian is the first project with Learning-to-Rank functionality added to it.

This seems quite a bold claim - for example, Terrier added support in 4.0 which looks like it was released in 2014.  Given we've not made a stable release of xapian-letor yet, I feel at best this might be technically correct, and I suspect in reality it's just not true.

File: advanced/learning_to_rank.rst [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+LTR can be broadly seen in two stages: Learning the model & Ranking. Learning the model takes the training file as input and produces a model. After that given this learnt model, when a new query comes in, scores can be assigned to the documents associated to it.
```
Comment (ojwb) [inline advanced/learning_to_rank.rst]: > Learning the model & Ranking

This isn't a character-limited medium like twitter or an SMS - say "and" not "&".

File: advanced/learning_to_rank.rst [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+2. Qrel file: This is the file containing relevance judgements. It should
+   be formatted in this way::
+
+    2010003 Q0 19243417 1
+    2010003 Q0 3256433 1
+    2010003 Q0 275014 1
+    2010003 Q0 298021 0
+    2010003 Q0 1456811 0
+
+   where first column is query-id, third column is Document-id and fourth column being relevance label which is 0 for irrelevance and 1 for relevance. Second column is many times referred as 'iter' but doesn't really important for us.  All the fields are whitespace delimited. This is the standard format of almost all the relevance judgement files. If you have little different relevance judgement file then you can easily convert it in such file using basic 'awk' command.
```
Comment (ojwb) [inline advanced/learning_to_rank.rst]: > If you have little different relevance judgement file then you can easily convert it in such file using basic 'awk' command.

Should be "convert it **to** such **a** file".

And "little different" sounds weird - I guess we're trying to say "If you have relevance judgements in a different format ..."

I'd also suggest thinking carefully about words like "simply", "easily", "trivial", etc in documentation.  Lana Brindley has a good talk about this I heard a few years back - I didn't find a video from a quick search, but this blog post covers some of the same ground:

http://lanabrindley.com/2011/11/20/open-source-docs-the-good-the-bad-and-the-wtf/

As a reader, if I don't see how to use "awk" to convert the format I have, then this sentence makes me feel stupid - this task is apparently "easy" but I've no idea how to do it, and now I'm stuck.  I may not know what "awk" is.

The word "easily" doesn't really add anything here.  Just dropping it is arguably an improvement, but we can also briefly explain what "awk" is - "... then you can convert it to this format using a text-processing tool such as 'awk'".  That's more helpful for a reader who hasn't heard of "awk" but has a favourite text-processing tool.

File: advanced/learning_to_rank.rst [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+3. Collection Index : Here you supply the path to the index of the corpus. If
+   you have 'title' information in the collection with some xml/html tag or so
+   then add::
+
+    indexer.index(title,1,"S");
+
+You can refer to the "Indexing" section under "A practical example" heading for the Collection Index. The database created in the practical example will be used as the collection index for the examples. In particular we are going to be using all the documents from which contain the term "watch" which will be used as the query for the examples.
```
Comment (ojwb) [inline advanced/learning_to_rank.rst]: _all the documents which contain the term "watch", which will be used as the query for the examples._ (lose a _from_, add a comma for clarity)

File: advanced/learning_to_rank.rst [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+Provided such information, API is capable of creating the training file which is in the mentioned format and can be easily used for learning a model.
```
Comment (ojwb) [inline advanced/learning_to_rank.rst]: Another "easily" which doesn't really add anything.

File: advanced/learning_to_rank.rst [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+You can use any one of the rankers to Learn the model. The command line tool xapian-train uses ListNET as the ranker for learning. To learn a model run the following command from the top level directory.
+
+.. code-block:: none
+
+    $ xapian-train --db=db data/training_data.txt "ListNET_Ranker"
+
+Ranking
+-------
+
+After we have built a model, its quite straightforward to get a real score for a particular document for the given query. Here we supply the first hand retrieved ranked-list to the Ranking function, which assigns a new score to each document after converting it to the same dimensioned feature vector. This list is re-ranked according to the new scores.
```
Comment (ojwb) [inline advanced/learning_to_rank.rst]: > its quite straightforward

Should be "it's" (since it's a contraction of "it is").

File: conf.py [hunk 7]
```diff
@@ -319,8 +319,12 @@ def xapian_code_example_command(ex):
     elif highlight_language == 'tcl':
         return "tclsh %s" % xapian_code_example_filename(ex)
     elif highlight_language == 'c++':
-        return "g++ `xapian-config --cxxflags` %s code/c++/support.cc -o %s `xapian-config --libs`\n./%s" \
-            % (xapian_code_example_filename(ex), ex, ex)
+        if option == 0:
+            return "g++ `xapian-config --cxxflags` %s code/c++/support.cc  -o %s `xapian-config --libs`\n./%s" \
+                % (xapian_code_example_filename(ex), ex, ex)
```
Comment (ojwb) [inline conf.py]: You seem to have introduced an extra space before "-o" here (and then copied it below).

File: conf.py [hunk 8]
```diff
@@ -375,8 +379,12 @@ def xapian_run_example_command(ex):
             if os.system('%s --libs > /dev/null 2>&1' % xapian_config) != 0:
                 pfx = 'libtool --quiet --mode=link '
                 xcopt = '--ltlibs'
-        return "%s%s `%s --cxxflags` %s code/c++/support.cc -o code/c++/built/%s `%s %s`\ncode/c++/built/%s" \
-            % (pfx, cxx, xapian_config, xapian_code_example_filename(ex), ex, xapian_config, xcopt, ex)
+        if option == 0:
+            return "%s%s `%s --cxxflags` %s code/c++/support.cc  -o code/c++/built/%s `%s %s`\ncode/c++/built/%s" \
+                % (pfx, cxx, xapian_config, xapian_code_example_filename(ex), ex, xapian_config, xcopt, ex)
```
Comment (ojwb) [inline conf.py]: Extra space again.

File: conf.py [hunk 9]
```diff
@@ -538,6 +550,15 @@ def run(self):
...
 
+class XapianTrain(LiteralInclude):
+    option_spec = { }
+
+    def run(self):
+        os.system("xapian-train --db=db data/training_data.txt ListNET_Ranker")
+        return super(XapianTrain, self).run()
+
+
+directives.register_directive('xapiantrain', XapianTrain)
```
Comment (ojwb) [inline conf.py]: Extra blank line before this one.

Also, for the other directives we have a "Usage:" comment.

File: conf.py [hunk 10]
```diff
@@ -303,7 +303,7 @@ def xapian_code_example_filename(ex):
     return "code/%s/%s%s" % (highlight_language, ex, ext)
 
 # Return the command to show in the generated docs.
-def xapian_code_example_command(ex):
+def xapian_code_example_command(ex, option):
```
Comment (ojwb) [inline conf.py]: `option` with values `0` or `1` is rather cryptic.  Why not call the parameter `letor`, and make the values `False` and `True`?  Or pass in the `options` dict.

File: code/c++/search_letor.cc [hunk 11]
```diff
@@ -0,0 +1,94 @@
+#include <xapian-letor.h>
```
Comment (ojwb) [inline code/c++/search_letor.cc]: This file seems to be heavily based on `xapian/xapian-letor/bin/xapian-rank.cc`, but you've stripped off the copyright statements and licensing boilerplate.  You really must not do that.

File: code/c++/search_letor.cc [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    "a", "about", "an", "and", "are", "as", "at",
+    "be", "by",
+    "en",
+    "for", "from",
+    "how",
+    "i", "in", "is", "it",
+    "of", "on", "or",
+    "that", "the", "this", "to",
+    "was", "what", "when", "where", "which", "who", "why", "will", "with"
+};
```
Comment (ojwb) [inline code/c++/search_letor.cc]: IIRC, James said **NOT** to copy the example code from xapian-letor.

But if you're going to ignore that, you need to simplify it by removing stuff which doesn't illustrate what the documentation is trying to explain.

For example, this stopword stuff just complicates the code such that it's harder to follow the letor part.  Similarly, spelling correction just gets in the way here.

File: code/c++/search_letor.cc [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    // Initialise Ranker object with ListNETRanker instance, db path and query.
+    // See Ranker documentation for available Ranker subclass options.
+    Xapian::ListNETRanker ranker;
+    ranker.set_database_path(db_path);
+    ranker.set_query(query);
+
+    // Get vector of re-ranked docids
+    ranker.rank(mset, model_key);
```
Comment (ojwb) [inline code/c++/search_letor.cc]: This comment is out-of-date now (`mset` gets reranked, there's no "vector" now).  That's also wrong in `xapian-rank.cc`.

File: code/c++/search_letor.cc [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+	cout << *i << ": [" << i.get_weight() << "]\n" << data << "\n";
+    }
+}
+// End of example code.
+
+int main(int argc, char** argv)
+{
+    if (argc != 4) {
+	show_usage();
+	return 0;
```
Comment (ojwb) [inline code/c++/search_letor.cc]: This is reporting that an incorrect number of arguments were passed - that's an error, so the program shouldn't exit with status 0.

File: code/c++/search_letor.cc [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+#include <iostream>
+#include <sstream>
+#include <string>
+
+using namespace std;
+
+static void show_usage()
+{
+    cout << "Usage: rank_letor --db=DIRECTORY MODEL_METADATA_KEY QUERY\n";
```
Comment (ojwb) [inline code/c++/search_letor.cc]: *  `rank_letor` doesn't match `search_letor.cc`
* The code actually expects just a database path for the first parameter, not with `--db=`
* `DIRECTORY` is inaccurate - a Xapian database can be a single file (glass supports a read-only single file variant, and there's also stub database files).

**Comment by AyushP123** [comment]
I have published the c++ documentation online [here](https://ayushp123.github.io/). Learning to Rank is published [here](https://ayushp123.github.io/advanced/learning_to_rank.html). We can use this site to provide c++ documentation along with the default python documentation that we have.

File: code/c++/search_letor.cc [hunk 16]
```diff
@@ -8,26 +29,12 @@ using namespace std;
 
 static void show_usage()
 {
-    cout << "Usage: rank_letor --db=DIRECTORY MODEL_METADATA_KEY QUERY\n";
+    cout << "Usage: search_letor --db=DBPATH MODEL_METADATA_KEY QUERY\n";
```
Comment (ojwb) [inline code/c++/search_letor.cc]: `--db=` isn't expected here.

File: conf.py [hunk 17]
```diff
@@ -319,11 +319,11 @@ def xapian_code_example_command(ex, option):
     elif highlight_language == 'tcl':
         return "tclsh %s" % xapian_code_example_filename(ex)
     elif highlight_language == 'c++':
-        if option == 0:
-            return "g++ `xapian-config --cxxflags` %s code/c++/support.cc  -o %s `xapian-config --libs`\n./%s" \
+        if letor == False:
```
Comment (ojwb) [inline conf.py]: It's generally regarded as better style not to compare boolean values against boolean literals, so `if not letor:` instead of `if letor == False:`.

(It's particularly problematic in languages without a real bool type, like C pre C99 - people would often make their own with something like `#define FALSE 0` and `#define TRUE 1` and `typedef int bool;` - problem is that a `bool` variable could end up with a value other than `0` or `1`, which is then true but not `TRUE` - `if (b)` is taken, but `if (b == TRUE)` isn't).

File: conf.py [hunk 18]
```diff
@@ -379,11 +379,11 @@ def xapian_run_example_command(ex, option):
             if os.system('%s --libs > /dev/null 2>&1' % xapian_config) != 0:
                 pfx = 'libtool --quiet --mode=link '
                 xcopt = '--ltlibs'
-        if option == 0:
-            return "%s%s `%s --cxxflags` %s code/c++/support.cc  -o code/c++/built/%s `%s %s`\ncode/c++/built/%s" \
+        if letor == False:
```
Comment (ojwb) [inline conf.py]: Ditto.

**Review by ojwb — CHANGES_REQUESTED** [review]
Thanks for the update - a few minor things:

**Comment by AyushP123** [comment]
Hi,
     I am very sorry for the late fixup. There are more issues present in the documentation which need to be ironed out. I remember addy getting confused on a lot of topics. Will go through them again and create a new PR.

**Comment by VaibhavKansagara** [comment]
@AyushP123 let me know if I can help with any of those issues.

**Comment by ojwb** [comment]
> Will go through them again and create a new PR.

I'd suggest it's probably worth addressing the last few issues here first so we can merge this first as it's already a significant improvement over the current situation.

Then a new PR can build on that.

File: code/c++/search_letor.cc (hunk 16 above)
Comment (VaibhavKansagara) [reply code/c++/search_letor.cc]: I think db_path is expected here.

File: code/c++/search_letor.cc [hunk 19]
```diff
@@ -29,7 +29,7 @@ using namespace std;
 
 static void show_usage()
 {
-    cout << "Usage: search_letor --db=DBPATH MODEL_METADATA_KEY QUERY\n";
+    cout << "Usage: search_letor MODEL_METADATA_KEY QUERY\n";
```
Comment (ojwb) [inline code/c++/search_letor.cc]: This change is wrong - it's **only** the `--db=` part which isn't expected - the `DBPATH` part is.

File: code/c++/search_letor.cc (hunk 16 above)
Comment (ojwb) [reply code/c++/search_letor.cc]: Indeed, this wasn't addressed correctly - I've commented on the fixup commit.

File: advanced/learning_to_rank.rst [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+    2010003 Q0 298021 0
+    2010003 Q0 1456811 0
+
+   where first column is query-id, third column is Document-id and fourth column being relevance label which is 0 for irrelevance and 1 for relevance. Second column is many times referred as 'iter' but doesn't really important for us.  All the fields are whitespace delimited. This is the standard format of almost all the relevance judgement files. If you have relevance judgements in a different format then you can convert it to this format using a text processing tool such as 'awk'.
+
+3. Collection Index : Here you supply the path to the index of the corpus. If
+   you have 'title' information in the collection with some xml/html tag or so
+   then add::
+
+    indexer.index(title,1,"S");
```
Comment (ojwb) [inline advanced/learning_to_rank.rst]: Putting spaces after the commas in code would better match the other code examples.
